### DISCUSSION TITLE
How to fix reset infinite loop while initializing USB PHY on STM32H750VB?

### DISCUSSION THREAD

**Discussion by bkaz4283** [hathach/tinyusb#769] [Q&A] [answered]
Hello,

I'm developing firmware for STM32H750VB that uses the USB PHY to send MIDI over USB. I'm using the TinyUSB library for the USB stack. Right now I'm having trouble initializing the USB PHY due to it getting stuck in an infinite loop because it's not clearing the reset signal. In other words, the code is getting stuck in this while loop:

usb_otg->GRSTCTL |= USB_OTG_GRSTCTL_CSRST;
while ((usb_otg->GRSTCTL & USB_OTG_GRSTCTL_CSRST) == USB_OTG_GRSTCTL_CSRST) {}

I checked that USB is getting the 48MHz clock it should be getting and in the RCC_AHB1ENR register, both USB2OTGFSEN and USB2OTGFSULPIEN bits are set.

I checked USB33RDY as well and it's zero all the time. Do I need to power on the 3.3v USB I/O supply in order for it to clear its reset? If so, I'm confused about how to even do this because the 100-pin LQFP package doesn't even have the VDD33USB/VDD50USB pins.

My product is a device only and is self-powered.

Can you think of anything else I should check that might be preventing the USB PHY from coming out of reset?

Thanks!
-Brian

**Comment by hathach** [comment]
1. does H750 has integrated or external HS PHY
2. are you using OTG HS port, if you what is your tusb_config.h

**Reply by bkaz4283** [reply]
1) Yes, H750 has integrated HS PHY.  But I only need it in FS mode.
2) I'm using the USB2 OTG_HS port (OTG_HS2) as this is the only one that supports FS mode.  I'm using the tusb_config.h from the midi_test folder in examples.  I added
#define CFG_TUSB_MCU	OPT_MCU_STM32H7
at the beginning of tusb_config.h but I think it has no effect since it doesn't come up in any of the cases handled there.  The only thing I can see really happening in tusb_config.h is that these two macros are defined:
#define CFG_TUSB_OS               OPT_OS_NONE
#define CFG_TUD_MIDI              1

**Reply by hathach** [reply]
make sure you pick the correct RHPORT in tusb_config.h along with its support speed. You could look up the bsp for H7 https://github.com/hathach/tinyusb/tree/master/hw/bsp/stm32h7 also checkout the board config that have similar setup to you as reference.

**Reply by bkaz4283** [reply]
OK, I see that BOARD_DEVICE_RHPORT_NUM can be defined as either 0 or 1.  Your examples folder doesn't contain anything using STM32H750x so I'm not sure what it should be in my case.

However, BOARD_DEVICE_RHPORT_NUM defaults to 0 and that didn't work so I tried 1.  Unfortunately, that didn't work either (hardware STILL stuck in the reset loop).

**Reply by hathach** [reply]
I don't have any H750 board so it is as far as I could suggest. maybe you should ask on st forum, people there could help you.

**Reply by bkaz4283** [reply]
Of course I opened a query with ST technical support in parallel with writing on this forum in case they might be able to help too.  You shouldn't be surprised to know that they advised me NOT to use TinyUSB, as they do not support it, and recommended using their HAL-based code produced by their STM32CubeMX tool.

Well, I was attracted to TinyUSB precisely because it does not use these bloated HAL libraries and it seemed the code could be much more compact and elegant.  It will be a huge disappointment if we cannot get USB to initialize properly because of some weird hardware configuration issue that we cannot resolve due to some probably obscure differences between H743/5 and H750!

**Reply by hathach** [reply]
Have you tried with one of supported H743 board, it should work with most if not all existing examples.

**Comment by bkaz4283** [comment, marked as answer]
I found the solution with the help of ST technical support.

Apparently the USB2OTGFSEN bit in the RCC_AHB1ENR register is supposed to be set **but the USB2OTGFSULPIEN should NOT be set.**

This is of course not adequately explained in ST's reference manual like countless other things.  Maybe if someone finds this thread here it will save them some of the time I wasted tracking this issue down...

**Reply by hathach** [reply]
does this need any changes in the dcd_init() of the stack ? If so that would be great if you could make an PR to help others in the future.

**Reply by bkaz4283** [reply]
I'm still struggling with getting my own code to work.  If I encounter more issues in the future I'll write about them in the discussion forum.  I'm not ready yet to take responsibility for changing open source code!

**Reply by hathach** [reply]
no problem at all, just paste the code here if you feel more comfortable. Once I have an H750, I will test it out.
